File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func balanceBST(_ root: TreeNode?) -> TreeNode? {
        var sortedValues: [Int] = []
        inOrderTraversal(root , &sortedValues);
        return constructBST(sortedValues, 0, sortedValues.count - 1)
    }
    private func inOrderTraversal(_ node: TreeNode? , _ result: inout[Int]){
        guard let node = node else { return }
        inOrderTraversal(node.left , &result)
        result.append(node.val)
        inOrderTraversal(node.right , &result)
    }

    private func constructBST(_ values: [Int] , _ start: Int , _ end: Int) -> TreeNode? {
        if start > end {
            return nil
        }

        let mid = start + (end - start) / 2
        let node = TreeNode(values[mid])
        node.left = constructBST(values , start , mid - 1)
        node.right = constructBST(values , mid + 1 , end)
        return node
    }

}

File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* balanceBST(TreeNode* root) {
       vector<int> ans;
       inOrderTraversal(root , ans);
       return constructBalancedBST(ans , 0 , ans.size() - 1);
    }
private:
    void inOrderTraversal(TreeNode* node , vector<int>& ans){
        if(node == NULL) return;
        inOrderTraversal(node->left , ans);
        ans.push_back(node->val);
        inOrderTraversal(node->right , ans);
    }

    TreeNode* constructBalancedBST(vector<int>& ans , int start , int end){
        if(start > end) return NULL;
        int mid = start + (end - start) / 2;
        TreeNode* node = new TreeNode(ans[mid]);
        node->left = constructBalancedBST(ans , start , mid - 1);
        node->right = constructBalancedBST(ans , mid+1 , end);
        return node;
    }
};